File: runtime/runtime.cpp
// Copyright (C) 2018 Tomoyuki Fujimori <moyu@dromozoa.com>
//
// This file is part of dromozoa-compiler.
//
// dromozoa-compiler is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// dromozoa-compiler is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with dromozoa-compiler.  If not, see <http://www.gnu.org/licenses/>.

#include <cctype>
#include <cmath>
#include <exception>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <utility>

#include "runtime.hpp"

namespace dromozoa {
  namespace runtime {
    namespace {
      template <class T>
      inline value_t* copy(value_t* result, const T& range) {
        return std::copy(range.begin(), range.end(), result);
      }

      inline value_t* copy(value_t* result, const value_t& value) {
        *result = value;
        ++result;
        return result;
      }

      inline value_t* copy(value_t* result, const value_t* begin, const value_t* end) {
        return std::copy(begin, end, result);
      }

      inline int regexp_integer(const std::string& string) {
        int state = 6;
        for (const char c : string) {
          switch (state) {
            case 1: if      ( std::isspace(c))      { state = 2; }
                    else if ( isdigit(c))           { state = 3; }
                    else if ( c == 'X' || c == 'x') { state = 8; }
                    else                            { return  0; } break;
            case 2: if      (!std::isspace(c))      { return  0; } break;
            case 3: if      ( std::isspace(c))      { state = 2; }
                    else if (!std::isdigit(c))      { return  0; } break;
            case 4: if      ( std::isspace(c))      { state = 5; }
                    else if (!std::isxdigit(c))     { return  0; } break;
            case 5: if      (!std::isspace(c))      { return  0; } break;
            case 6: if      ( c == '0')             { state = 1; }
                    else if ( std::isdigit(c))      { state = 3; }
                    else if ( c == '+' || c == '-') { state = 7; }
                    else if (!std::isspace(c))      { return  0; } break;
            case 7: if      ( c == '0')             { state = 1; }
                    else if ( std::isdigit(c))      { state = 3; }
                    else                            { return  0; } break;
            case 8: if      ( std::isxdigit(c))     { state = 4; }
                    else                            { return  0; } break;
          }
        }
        switch (state) {
          case 1: case 2: case 3: return 1;
          case 4: case 5:         return 2;
          default:                return 0;
        }
      }

      inline bool to_number(const std::string& string, double& result) {
        try {
          const auto n = string.find_last_not_of(" \f\n\r\t\v");
          if (n != std::string::npos) {
            std::size_t i = 0;
            const auto number = std::stod(string, &i);
            if (i == n + 1) {
              result = number;
              return true;
            }
          }
        } catch (const std::exception&) {}
        return false;
      }

      inline std::string to_string(double number) {
        std::ostringstream out;
        out << std::setprecision(17) << number;
        return out.str();
      }
    }

    struct value_t::access {
      static void copy_construct(value_t& self, const value_t& that) {
        self.type_ = that.type_;
        switch (self.type_) {
          case type_t::nil:
            return;
          case type_t::boolean:
            self.boolean_ = that.boolean_;
            return;
          case type_t::number:
            self.number_ = that.number_;
            return;
          case type_t::string:
            new (&self.string_) std::shared_ptr<std::string>(that.string_);
            return;
          case type_t::table:
            new (&self.table_) std::shared_ptr<table_t>(that.table_);
            return;
          case type_t::function:
            new (&self.function_) std::shared_ptr<function_t>(that.function_);
            return;
          case type_t::thread:
            new (&self.thread_) std::shared_ptr<thread_t>(that.thread_);
            return;
        }
      }

      static void move_construct(value_t& self, value_t&& that) {
        self.type_ = that.type_;
        that.type_ = type_t::nil;
        switch (self.type_) {
          case type_t::nil:
            return;
          case type_t::boolean:
            self.boolean_ = that.boolean_;
            return;
          case type_t::number:
            self.number_ = that.number_;
            return;
          case type_t::string:
            new (&self.string_) std::shared_ptr<std::string>(std::move(that.string_));
            return;
          case type_t::table:
            new (&self.table_) std::shared_ptr<table_t>(std::move(that.table_));
            return;
          case type_t::function:
            new (&self.function_) std::shared_ptr<function_t>(std::move(that.function_));
            return;
          case type_t::thread:
            new (&self.thread_) std::shared_ptr<thread_t>(std::move(that.thread_));
            return;
        }
      }

      static void destruct(value_t& self) {
        switch (self.type_) {
          case type_t::nil:
            return;
          case type_t::boolean:
            return;
          case type_t::number:
            return;
          case type_t::string:
            self.string_.~shared_ptr();
            return;
          case type_t::table:
            self.table_.~shared_ptr();
            return;
          case type_t::function:
            self.function_.~shared_ptr();
            return;
          case type_t::thread:
            self.thread_.~shared_ptr();
            return;
        }
      }
    };

    array_t::array_t()
      : size_(0) {}

    array_t::array_t(std::initializer_list<value_t> data)
      : size_(data.size()) {
      if (size_ > 0) {
        data_ = std::shared_ptr<value_t>(new value_t[size_], std::default_delete<value_t[]>());
        copy(data_.get(), data);
      }
    }

    array_t::array_t(std::initializer_list<value_t> data, const array_t& that)
      : size_(data.size() + that.size()) {
      if (size_ > 0) {
        data_ = std::shared_ptr<value_t>(new value_t[size_], std::default_delete<value_t[]>());
        copy(copy(data_.get(), data), that);
      }
    }

    array_t::array_t(const value_t& value, const array_t& that)
      : size_(1 + that.size()) {
      data_ = std::shared_ptr<value_t>(new value_t[size_], std::default_delete<value_t[]>());
      copy(copy(data_.get(), value), that);
    }

    const value_t* array_t::begin() const {
      return data_.get();
    }

    const value_t* array_t::end() const {
      return data_.get() + size_;
    }

    const value_t& array_t::operator[](std::size_t i) const {
      if (i < size_) {
        return data_.get()[i];
      } else {
        return NIL;
      }
    }

    std::size_t array_t::size() const {
      return size_;
    }

    array_t array_t::slice(std::size_t first) const {
      if (size_ > first) {
        array_t that;
        that.size_ = size_ - first;
        that.data_ = std::shared_ptr<value_t>(new value_t[that.size_], std::default_delete<value_t[]>());
        copy(that.data_.get(), begin() + first, end());
        return that;
      } else {
        return {};
      }
    }

    state_t::state_t()
      : string_metatable_(std::make_shared<table_t>()),
        env_(std::make_shared<table_t>()) {}

    std::shared_ptr<table_t> state_t::string_metatable() const {
      return string_metatable_;
    };

    std::shared_ptr<table_t> state_t::env() const {
      return env_;
    };

    void state_t::open_base() {
      env_->set("tostring", make_function([](continuation_t k, state_t state, array_t args) {
        return k(state, { "-" });
      }));
    }

    void state_t::open_io() {
      value_t module { std::make_shared<table_t>() };

      module.rawset("write", make_function([](continuation_t k, state_t state, array_t args) {
        for (const auto& value : args) {
          std::cout << value.checkstring();
        }
        return k(state, {});
      }));

      env_->set("io", module);
    }

    void state_t::open_string() {
    }

    value_t::value_t()
      : type_(type_t::nil) {}

    value_t::value_t(const value_t& that)
      : value_t() {
      access::copy_construct(*this, that);
    }

    value_t::value_t(value_t&& that)
      : value_t() {
      access::move_construct(*this, std::move(that));
    }

    value_t::~value_t() {
      access::destruct(*this);
    }

    value_t& value_t::operator=(const value_t& that) {
      access::destruct(*this);
      access::copy_construct(*this, that);
      return *this;
    }

    value_t& value_t::operator=(value_t&& that) {
      access::destruct(*this);
      access::move_construct(*this, std::move(that));
      return *this;
    }

    value_t::value_t(bool boolean)
      : type_(type_t::boolean),
        boolean_(boolean) {}

    value_t::value_t(double number)
      : type_(type_t::number),
        number_(number) {}

    value_t::value_t(const char* data)
      : type_(type_t::string) {
      new (&string_) std::shared_ptr<std::string>(std::make_shared<std::string>(data));
    }

    value_t::value_t(const char* data, std::size_t size)
      : type_(type_t::string) {
      new (&string_) std::shared_ptr<std::string>(std::make_shared<std::string>(data, size));
    }

    value_t::value_t(const std::string& string)
      : type_(type_t::string) {
      new (&string_) std::shared_ptr<std::string>(std::make_shared<std::string>(string));
    }

    value_t::value_t(std::string&& string)
      : type_(type_t::string) {
      new (&string_) std::shared_ptr<std::string>(std::make_shared<std::string>(std::move(string)));
    }

    value_t::value_t(std::shared_ptr<table_t> table)
      : type_(type_t::table) {
      new (&table_) std::shared_ptr<table_t>(table);
    }

    value_t::value_t(std::shared_ptr<function_t> function)
      : type_(type_t::function) {
      new (&function_) std::shared_ptr<function_t>(function);
    }

    value_t::value_t(std::shared_ptr<thread_t> thread)
      : type_(type_t::thread) {
      new (&thread_) std::shared_ptr<thread_t>(thread);
    }

    bool value_t::operator<(const value_t& that) const {
      if (type_ != that.type_) {
        return type_ < that.type_;
      }
      switch (type_) {
        case type_t::nil:
          return false;
        case type_t::boolean:
          return boolean_ < that.boolean_;
        case type_t::number:
          return number_ < that.number_;
        case type_t::string:
          return *string_ < *that.string_;
        case type_t::table:
          return table_ < that.table_;
        case type_t::function:
          return function_ < that.function_;
        case type_t::thread:
          return thread_ < that.thread_;
      }
    }

    std::string value_t::type() const {
      switch (type_) {
        case type_t::nil:
          return "nil";
        case type_t::boolean:
          return "boolean";
        case type_t::number:
          return "number";
        case type_t::string:
          return "string";
        case type_t::table:
          return "table";
        case type_t::function:
          return "function";
        case type_t::thread:
          return "thread";
      }
    }

    bool value_t::isnil() const {
      return type_ == type_t::nil;
    }

    bool value_t::isboolean() const {
      return type_ == type_t::boolean;
    }

    bool value_t::isnumber() const {
      return type_ == type_t::number;
    }

    bool value_t::isstring() const {
      return type_ == type_t::string;
    }

    bool value_t::istable() const {
      return type_ == type_t::table;
    }

    bool value_t::isfunction() const {
      return type_ == type_t::function;
    }

    bool value_t::isthread() const {
      return type_ == type_t::thread;
    }

    bool value_t::toboolean() const {
      if (isnil()) {
        return false;
      } else if (isboolean()) {
        return boolean_;
      } else {
        return true;
      }
    }

    bool value_t::tonumber(double& result) const {
      if (isnumber()) {
        result = number_;
        return true;
      } else if (isstring()) {
        switch (regexp_integer(*string_)) {
          case 1: result = std::stoll(*string_, nullptr, 10); return true;
          case 2: result = std::stoll(*string_, nullptr, 16); return true;
        }
        double number = 0;
        if (to_number(*string_, number)) {
          result = number;
          return true;
        }
      }
      return false;
    }

    double value_t::checknumber() const {
      double result = 0;
      if (tonumber(result)) {
        return result;
      }
      throw value_t("number expected, got " + type());
    }

    std::int64_t value_t::checkinteger() const {
      if (isnumber()) {
        if (std::isfinite(number_) && number_ == std::floor(number_)) {
          return number_;
        }
        throw value_t("number has no integer representation");
      } else if (isstring()) {
        switch (regexp_integer(*string_)) {
          case 1: return std::stoll(*string_, nullptr, 10);
          case 2: return std::stoll(*string_, nullptr, 16);
        }
        double number = 0;
        if (to_number(*string_, number)) {
          if (std::isfinite(number) && number == std::floor(number)) {
            return number;
          }
          throw value_t("number has no integer representation");
        }
      }
      throw value_t("integer expected, got " + type());
    }

    std::string value_t::checkstring() const {
      if (isstring()) {
        return *string_;
      } else if (isnumber()) {
        return to_string(number_);
      }
      throw value_t("string expected, got " + type());
    }

    std::shared_ptr<table_t> value_t::checktable() const {
      if (istable()) {
        return table_;
      }
      throw value_t("table expected, got " + type());
    }

    std::shared_ptr<function_t> value_t::checkfunction() const {
      if (isfunction()) {
        return function_;
      }
      throw value_t("function expected, got " + type());
    }

    std::shared_ptr<thread_t> value_t::checkthread() const {
      if (isthread()) {
        return thread_;
      }
      throw value_t("thread expected, got " + type());
    }

    std::int64_t value_t::optinteger(std::int64_t d) const {
      if (isnil()) {
        return d;
      } else {
        return checkinteger();
      }
    }

    std::int64_t value_t::rawlen() const {
      if (isstring()) {
        return string_->size();
      } else if (istable()) {
        for (std::int64_t i = 1; ; ++i) {
          if (rawget(i).isnil()) {
            return i - 1;
          }
        }
      }
      throw value_t("attempt to get length of a " + type() + " value");
    }

    const value_t& value_t::rawget(const value_t& index) const {
      return checktable()->get(index);
    }

    void value_t::rawset(const value_t& index, const value_t& value) const {
      checktable()->set(index, value);
    }

    void value_t::rawset(const value_t& begin, const array_t& array) const {
      const auto table = checktable();
      std::int64_t index = begin.checkinteger();
      for (std::size_t i = 0; i < array.size(); ++i) {
        table->set(index++, array[i]);
      }
    }

    const value_t& value_t::getmetafield(const state_t& state, const value_t& event) const {
      if (isstring()) {
        return state.string_metatable()->get(event);
      } else if (istable()) {
        const auto metatable = table_->getmetatable();
        if (metatable.istable()) {
          return metatable.rawget(event);
        }
      }
      return NIL;
    }

    std::shared_ptr<thunk_t> value_t::call(continuation_t k, state_t state, array_t args) const {
      if (isfunction()) {
        return (*checkfunction())(k, state, args);
      } else {
        const auto& field = getmetafield(state, "__call");
        if (field.isfunction()) {
          return (*field.checkfunction())(k, state, array_t(*this, args));
        }
      }
      throw value_t("attempt to call a " + type() + " value");
    }

    std::shared_ptr<thunk_t> value_t::gettable(continuation_t k, state_t state, const value_t& index) const {
      if (istable()) {
        const auto& result = rawget(index);
        if (!result.isnil()) {
          return k(state, { result });
        }
      }
      const auto& field = getmetafield(state, "__index");
      if (field.isnil()) {
        return k(state, { NIL });
      } else if (field.isfunction()) {
        return field.call(k, state, { *this, index });
      } else {
        return field.gettable(k, state, index);
      }
    }

    const value_t NIL;
    const value_t FALSE = false;
    const value_t TRUE = true;

    var_t::var_t() {}

    var_t::var_t(const value_t& value)
      : value_(value) {}

    var_t::var_t(value_t&& value)
      : value_(std::move(value)) {}

    const value_t& var_t::operator*() const {
      return value_;
    }

    const value_t* var_t::operator->() const {
      return &value_;
    }

    ref_t::ref_t()
      : value_(std::make_shared<value_t>()) {}

    ref_t::ref_t(const value_t& value)
      : value_(std::make_shared<value_t>(value)) {}

    ref_t::ref_t(value_t&& value)
      : value_(std::make_shared<value_t>(std::move(value))) {}

    value_t& ref_t::operator*() const {
      return *value_;
    }

    value_t* ref_t::operator->() const {
      return value_.get();
    }

    const value_t& table_t::get(const value_t& index) const {
      const auto i = map_.find(index);
      if (i == map_.end()) {
        return NIL;
      } else {
        return i->second;
      }
    }

    void table_t::set(const value_t& index, const value_t& value) {
      if (index.isnil()) {
        throw value_t("table index is nil");
      }
      if (value.isnil()) {
        map_.erase(index);
      } else {
        map_[index] = value;
      }
    }

    const value_t& table_t::getmetatable() const {
      return metatable_;
    }

    void table_t::setmetatable(const value_t& metatable) {
      metatable_ = metatable;
    }

    thunk_t::~thunk_t() {}

    function_t::~function_t() {}
  }
}

File: runtime/runtime.js
// Copyright (C) 2018 Tomoyuki Fujimori <moyu@dromozoa.com>
//
// This file is part of dromozoa-compiler.
//
// dromozoa-compiler is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// dromozoa-compiler is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// Under Section 7 of GPL version 3, you are granted additional
// permissions described in the GCC Runtime Library Exception, version
// 3.1, as published by the Free Software Foundation.
//
// You should have received a copy of the GNU General Public License
// and a copy of the GCC Runtime Library Exception along with
// dromozoa-compiler.  If not, see <http://www.gnu.org/licenses/>.

const METATABLE = Symbol("metatabale");

const decint_pattern = /^\s*([+-]?\d+)\s*$/;
const hexint_pattern = /^\s*([+-]?0[xX][0-9A-Fa-f]+)\s*$/;
const decflt_pattern = /^\s*([+-]?(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?)\s*$/;

const string_buffers = new Map();
const string_metatable = new Map();

class logic_error {
  constructor(message) {
    this.message = message;
  }
}

class runtime_error {
  constructor(message) {
    this.message = message;
  }
}

const string_buffer = s => {
  let buffer = string_buffers.get(s);
  if (buffer !== undefined) {
    string_buffers.delete(s);
  } else {
    if (typeof TextEncoder !== "undefined") {
      buffer = new TextEncoder().encode(s);
    } else if (typeof Buffer !== "undefined") {
      buffer = Buffer.from(s);
    } else {
      throw new runtime_error("no UTF-8 encoder");
    }
  }
  string_buffers.set(s, buffer);
  if (string_buffers.size > 16) {
    for (const entry of string_buffers) {
      string_buffers.delete(entry[0]);
      break;
    }
  }
  return buffer;
};

const is_nil = (value) => {
  return value === undefined;
};

const is_boolean = (value) => {
  return typeof value === "boolean";
};

const is_number = (value) => {
  return typeof value === "number";
};

const is_string = (value) => {
  return typeof value === "string" || String.prototype.isPrototypeOf(value);
};

const is_table = (value) => {
  return Map.prototype.isPrototypeOf(value);
};

const is_function = (value) => {
  return typeof value === "function";
};

const type = value => {
  if (is_nil(value)) {
    return "nil";
  } else if (is_boolean(value)) {
    return "boolean";
  } else if (is_number(value)) {
    return "number";
  } else if (is_string(value)) {
    return "string";
  } else if (is_table(value)) {
    return "table";
  } else if (is_function(value)) {
    return "function";
  }
  throw new logic_error("unreachable code");
};

const toboolean = (v) => {
  if (is_nil(v)) {
    return false;
  } else if (is_boolean(v)) {
    return v;
  }
  return true;
};

const tonumber = v => {
  if (is_number(v)) {
    return v;
  } else if (is_string(v)) {
    let match;
    if ((match = decint_pattern.exec(v))) {
      return parseInt(match[0], 10);
    }
    if ((match = hexint_pattern.exec(v))) {
      return parseInt(match[0], 16);
    }
    if ((match = decflt_pattern.exec(v))) {
      return parseFloat(match[0]);
    }
  }
};


const checknumber = (v) => {
  const result = tonumber(v);
  if (is_number(result)) {
    return result;
  }
  throw new runtime_error("number expected, got " + type(v));
};

const checkinteger = (v) => {
  const result = checknumber(v);
  if (Number.isInteger(result)) {
    return result;
  }
  throw new runtime_error("number has no integer representation");
};

const checkstring = (v) => {
  if (is_string(v)) {
    return v;
  } else if (is_number(v)) {
    return v.toString();
  }
  throw new runtime_error("string expected, got " + type(v));
};

const checktable = (v) => {
  if (is_table(v)) {
    return v;
  }
  throw new runtime_error("table expected, got " + type(v));
};

const optinteger = (v, d) => {
  if (is_nil(v)) {
    return d;
  } else {
    return checkinteger(v);
  }
};

const rawget = (table, index) => {
  return checktable(table).get(index);
};

const rawset = (table, index, value) => {
  if (is_nil(index)) {
    throw new runtime_error("table index is nil");
  }
  if (is_nil(value)) {
    checktable(table).delete(index);
  } else {
    checktable(table).set(index, value);
  }
};

const getmetafield = (object, event) => {
  let metatable;
  if (is_string(object)) {
    metatable = string_metatable;
  } else if (is_table(object)) {
    metatable = object[METATABLE];
  }
  if (is_table(metatable)) {
    return rawget(metatable, event);
  }
};

const call0 = (f, ...args) => {
  if (is_function(f)) {
    f(...args);
  } else {
    const field = getmetafield(f, "__call");
    if (is_function(field)) {
      field(f, ...args);
    } else {
      throw new runtime_error("attempt to call a " + type(f) + " value");
    }
  }
};

const call1 = (f, ...args) => {
  let result;
  if (is_function(f)) {
    result = f(...args);
  } else {
    const field = getmetafield(f, "__call");
    if (is_function(field)) {
      result = field(f, ...args);
    } else {
      throw new runtime_error("attempt to call a " + type(f) + " value");
    }
  }
  if (Array.prototype.isPrototypeOf(result)) {
    return result[0];
  } else {
    return result;
  }
};

const call = (f, ...args) => {
  let result;
  if (is_function(f)) {
    result = f(...args);
  } else {
    const field = getmetafield(f, "__call");
    if (is_function(field)) {
      result = field(f, ...args);
    } else {
      throw new runtime_error("attempt to call a " + type(f) + " value");
    }
  }
  if (Array.prototype.isPrototypeOf(result)) {
    return result;
  } else {
    return [ result ];
  }
};

const getmetatable = object => {
  if (is_string(object)) {
    return string_metatable;
  } else if (is_table(object)) {
    const metatable = object[METATABLE];
    if (is_table(metatable)) {
      const protected_metatable = rawget(metatable, "__metatable");
      if (!is_nil(protected_metatable)) {
        return protected_metatable;
      }
    }
    return metatable;
  }
};

const setmetatable = (table, metatable) => {
  if (!is_nil(metatable) && !is_table(metatable)) {
    throw new runtime_error("nil or table expected");
  }
  if (!is_nil(getmetafield(table, "__metatable"))) {
    throw new runtime_error("cannot change a protected metatable");
  }
  checktable(table)[METATABLE] = metatable;
  return table;
};

const gettable = (table, index) => {
  if (is_string(table)) {
    const field = getmetafield(table, "__index");
    if (!is_nil(field)) {
      if (is_function(field)) {
        return call1(field, table, index);
      } else {
        return gettable(field, index);
      }
    }
  }
  const result = rawget(table, index);
  if (is_nil(result)) {
    const field = getmetafield(table, "__index");
    if (!is_nil(field)) {
      if (is_function(field)) {
        return call1(field, table, index);
      } else {
        return gettable(field, index);
      }
    }
  }
  return result;
};

const settable = (table, index, value) => {
  const result = rawget(table, index);
  if (is_nil(result)) {
    const field = getmetafield(table, "__newindex");
    if (!is_nil(field)) {
      if (is_function(field)) {
        return call0(field, table, index, value);
      } else {
        return settable(field, index, value);
      }
    }
  }
  rawset(table, index, value);
};

const setlist = (table, index, ...args) => {
  for (let i = 0; i < args.length; ++i) {
    rawset(table, index++, args[i]);
  }
};

const tostring = v => {
  if (is_nil(v)) {
    return "nil";
  } else if (is_boolean(v)) {
    if (v) {
      return "true";
    } else {
      return "false";
    }
  } else if (is_number(v)) {
    return v.toString();
  } else if (is_string(v)) {
    return v;
  } else if (is_table(v)) {
    const field = getmetafield(v, "__tostring");
    if (!is_nil(field)) {
      return call1(field, v);
    } else {
      return "table";
    }
  } else if (is_function(v)) {
    return "function";
  }
  throw logic_error("unreachable code");
};

const len = v => {
  if (is_string(v)) {
    return string_buffer(v).byteLength;
  } else if (is_table(v)) {
    const field = getmetafield(v, "__len");
    if (!is_nil(field)) {
      return call1(field, v);
    }
    for (let i = 1; ; ++i) {
      if (is_nil(gettable(v, i))) {
        return i - 1;
      }
    }
  }
  throw new runtime_error("attempt to get length of a " + type(v) + " value");
};

const eq = (self, that) => {
  if (self === that) {
    return true;
  }
  if (is_table(self) && is_table(that)) {
    let field = getmetafield(self, "__eq");
    if (is_nil(field)) {
      field = getmetafield(that, "__eq");
    }
    if (!is_nil(field)) {
      return toboolean(call1(field, self, that));
    }
  }
  return false;
};

const lt = (self, that) => {
  if (is_number(self) && is_number(that)) {
    return self < that;
  } else if (is_string(self) && is_string(that)) {
    return self < that;
  } else {
    let field = getmetafield(self, "__lt");
    if (is_nil(field)) {
      field = getmetafield(self, "__lt");
    }
    if (!is_nil(field)) {
      return toboolean(call1(field, self, that));
    }
  }
  throw new runtime_error("attempt to compare " + type(self) + " with " + type(that));
};

const le = (self, that) => {
  if (is_number(self) && is_number(that)) {
    return self <= that;
  } else if (is_string(self) && is_string(that)) {
    return self <= that;
  } else {
    let field = getmetafield(self, "__le");
    if (is_nil(field)) {
      field = getmetafield(self, "__le");
    }
    if (!is_nil(field)) {
      return toboolean(call1(field, self, that));
    }
    field = getmetafield(that, "__lt");
    if (is_nil(field)) {
      field = getmetafield(self, "__lt");
    }
    if (!is_nil(field)) {
      return !toboolean(call1(field, that, self));
    }
  }
  throw new runtime_error("attempt to compare " + type(self) + " with " + type(that));
};

const range_i = (i, size) => {
  if (i < 0) {
    i += size;
    if (i < 0) {
      return 0;
    } else {
      return i;
    }
  } else if (i > 0) {
    return i - 1;
  } else {
    return 0;
  }
};

const range_j = (j, size) => {
  if (j < 0) {
    j += size + 1;
    if (j < 0) {
      return 0;
    } else {
      return j;
    }
  } else if (j > size) {
    return size;
  } else {
    return j;
  }
};

const suppress_no_unsed = () => {};
suppress_no_unsed(len);
suppress_no_unsed(setlist);
suppress_no_unsed(lt);
suppress_no_unsed(le);

const open_base = env => {
  const ipairs_iterator = (table, index) => {
    index = checkinteger(index) + 1;
    const value = gettable(table, index);
    if (!is_nil(value)) {
      return [ index, value ];
    }
  };

  settable(env, "_G", env);

  settable(env, "_VERSION", "Lua 5.3");

  settable(env, "assert", (...args) => {
    const value = args[0];
    if (toboolean(value)) {
      return args;
    } else {
      if (args.length > 1) {
        throw new runtime_error(args[1]);
      } else {
        throw new runtime_error("assertion failed!");
      }
    }
  });

  settable(env, "error", message => {
    throw new runtime_error(message);
  });

  settable(env, "getmetatable", getmetatable);

  settable(env, "ipairs", table => {
    return [ ipairs_iterator, table, 0 ];
  });

  settable(env, "pcall", (f, ...args) => {
    try {
      const result = call(f, ...args);
      return [ true, ...result ];
    } catch (e) {
      if (runtime_error.prototype.isPrototypeOf(e)) {
        return [ false, e.message ];
      } else {
        throw e;
      }
    }
  });

  settable(env, "print", (...args) => {
    if (typeof process !== "undefined") {
      for (let i = 0; i < args.length; ++i) {
        if (i > 0) {
          process.stdout.write("\t");
        }
        process.stdout.write(tostring(args[i]));
      }
      process.stdout.write("\n");
    } else {
      console.log(...args);
    }
  });

  settable(env, "select", (index, ...args) => {
    if (eq(index, "#")) {
      return args.length;
    }
    const min = range_i(checkinteger(index), args.length);
    const result = [];
    for (let i = min; i < args.length; ++i) {
      result.push(args[i]);
    }
    return result;
  });

  settable(env, "setmetatable", setmetatable);

  settable(env, "tonumber", tonumber);

  settable(env, "tostring", tostring);

  settable(env, "type", type);
};

const open_string = env => {
  const module = new Map();

  settable(module, "byte", (s, i, j) => {
    const buffer = string_buffer(s);
    const index = optinteger(i, 1);
    const min = range_i(index, buffer.byteLength);
    const max = range_j(optinteger(j, index), buffer.byteLength);
    const result = [];
    for (let i = min; i < max; ++i) {
      result.push(buffer[i]);
    }
    return result;
  });

  settable(module, "char", (...args) => {
    if (typeof TextDecoder !== "undefined") {
      return new TextDecoder().decode(new Uint8Array(args));
    } else if (typeof Buffer !== "undefined") {
      return Buffer.from(args).toString();
    } else {
      throw new runtime_error("no UTF-8 decoder");
    }
  });

  settable(module, "len", s => {
    return string_buffer(s).byteLength;
  });

  settable(module, "sub", (s, i, j) => {
    const buffer = string_buffer(checkstring(s));
    const min = range_i(optinteger(i, 1), buffer.byteLength);
    const max = range_j(optinteger(j, buffer.byteLength), buffer.byteLength);
    if (min < max) {
      if (typeof TextDecoder !== "undefined") {
        return new TextDecoder().decode(buffer.slice(min, max));
      } else if (typeof Buffer !== "undefined") {
        return buffer.slice(min, max).toString();
      } else {
        throw new runtime_error("no UTF-8 decoder");
      }
    } else {
      return "";
    }
  });

  settable(env, "string", module);
  string_metatable.set("__index", module);
};

const env = new Map();
open_base(env);
open_string(env);
